core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -922,60 +1363,180 @@ namespace kagome::consensus::grandpa {
-                j.items.push_back(equivocatory_voting_message.second);
+              [&justification](const EquivocatoryVotingMessage
+                                   &equivocatory_voting_message) {
+                justification.items.push_back(
+                    equivocatory_voting_message.first);
+                justification.items.push_back(
+                    equivocatory_voting_message.second);
               });
-          return j;
+          return justification;
         });
-    return justification;
+    return result;
+  }
+
+  void VotingRoundImpl::doCatchUpRequest(const libp2p::peer::PeerId &peer_id) {
+    //	TODO(xDimon): Perhaps need to check if peer is known validator
+    //
+    //	if (not voter_set_->voterIndex(peer_id).has_value()) {
+    //	  return;
+    //	}
+
+    auto res =
+        env_->onCatchUpRequested(peer_id, voter_set_->id(), round_number_);
+    if (not res) {
+      logger_->warn("Catch-Up-Request was not sent: {}", res.error().message());
+    }
+  }
+
+  void VotingRoundImpl::doCatchUpResponse(const libp2p::peer::PeerId &peer_id) {
+    BOOST_ASSERT(finalized_.has_value());
+    const auto &finalised_block = finalized_.value();
+
+    auto best_prevote_candidate = bestPrevoteCandidate();
+    auto prevote_justification_opt =
+        getJustification(best_prevote_candidate, prevotes_->getMessages());
+    BOOST_ASSERT(prevote_justification_opt.has_value());
+    auto &prevote_justification = prevote_justification_opt.value();
+
+    //    auto best_final_candidate = bestFinalCandidate();
+    auto precommit_justification_opt =
+        getJustification(finalised_block, precommits_->getMessages());
+    BOOST_ASSERT(precommit_justification_opt.has_value());
+    auto &precommit_justification = precommit_justification_opt.value();
+
+    auto result = env_->onCatchUpResponsed(peer_id,
+                                           voter_set_->id(),
+                                           round_number_,
+                                           std::move(prevote_justification),
+                                           std::move(precommit_justification),
+                                           finalised_block);
+    if (not result) {
+      logger_->warn("Catch-Up-Response was not sent: {}",
+                    result.error().message());
+    }
   }
 
+  void VotingRoundImpl::pending() {
+    pending_timer_.cancel();
+
+    // Resend defined votes
+    if (isPrimary_ && primary_vote_.has_value()) {
+      sendProposal(convertToPrimaryPropose(primary_vote_.value()));
+    }
+    if (prevote_.has_value()) {
+      sendPrevote(convertToPrevote(prevote_.value()));
+    }
+    if (precommit_.has_value()) {
+      sendPrecommit(convertToPrecommit(precommit_.value()));
+    }
+    attemptToFinalizeRound();
+
+    pending_timer_.expires_from_now(std::max<Clock::Duration>(
+        duration_ * 10,
+        std::chrono::seconds(15)));  // Should be longer than precommit timer
```

Comment:
It seems better to extract these constants to header:
```c++
static constexpr uint64_t precommit_timer = N;
static constexpr uint64_t pending_timer = precommit_timer + K;

Reply:
It already refactored by little bit different way.